File: programmers/dp/GPS.py
def solution(n, m, edge_list, k, gps_log):
    INF = float('inf')
    
    # 1. 그래프 구성 (인접 리스트)
    # n이 작으므로 dict보다 list가 접근 속도가 빠름
    graph = [[] for _ in range(n + 1)]
    for u, v in edge_list:
        graph[u].append(v)
        graph[v].append(u)
        
    # [Point] 제자리에 머무르는 경우(Self-loop) 추가
    for i in range(1, n + 1):
        graph[i].append(i)

    # 2. DP 초기화 (첫 번째 단계)
    # prev_dp[i]: 이전 시간 t-1에 i번 노드에 있을 최소 비용
    prev_dp = [INF] * (n + 1)
    start_node = gps_log[0]
    prev_dp[start_node] = 0

    # 3. DP 수행 (시간 t: 1 ~ k-1)
    for t in range(1, k):
        target = gps_log[t]
        curr_dp = [INF] * (n + 1)
        
        # 모든 노드(node)에 대해, 이전 단계에서 이 노드로 올 수 있는 경우를 확인
        for node in range(1, n + 1):
            # node로 연결된 인접 노드들(neighbor)의 이전 비용 중 최소값 찾기
            # (양방향 그래프이므로 graph[node]는 node로 들어올 수 있는 노드 목록과 같음)
            
            # Pythonic한 min 값 찾기:
            # 이전 단계(prev_dp)에서 node로 올 수 있는 인접 노드들의 비용들을 모음
            candidates = [prev_dp[neighbor] for neighbor in graph[node]]
            
            min_prev_cost = min(candidates)
            
            if min_prev_cost != INF:
                # 현재 위치(node)가 로그(target)와 다르면 비용 +1
                add_cost = 0 if node == target else 1
                curr_dp[node] = min_prev_cost + add_cost
        
        # 현재 단계를 이전 단계로 갱신 (Rolling Update)
        prev_dp = curr_dp

    # 4. 결과 반환
    # 마지막 시간(k-1)에 gps_log의 마지막 위치에 도달하는 비용
    end_node = gps_log[-1]
    result = prev_dp[end_node]
    
    return result if result != INF else -1

File: programmers/dp/GPS.cpp
#include <vector>
#include <algorithm>

using namespace std;

int solution(int n, int m, vector<vector<int>> edge_list, int k, vector<int> gps_log) {
    const int INF = 1e9;
    
    vector<vector<int>> graph(n + 1);
    
    for (auto &edge : edge_list) {
        int u = edge[0];
        int v = edge[1];
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    
    // 제자리에 머무를 수도 있음
    for (int i = 1; i <= n; i++) {
        graph[i].push_back(i);
    }
    
    vector<vector<int>> dp(k, vector<int>(n + 1, INF));
    
    // 시작점은 gps_log[0]으로 고정
    dp[0][gps_log[0]] = 0;
    
    for (int t = 1; t < k; t++) {
        for (int prev = 1; prev <= n; prev++) {
            if (dp[t - 1][prev] == INF) continue;
            
            for (int next : graph[prev]) {
                int cost = (next == gps_log[t]) ? 0 : 1;
                dp[t][next] = min(dp[t][next], dp[t - 1][prev] + cost);
            }
        }
    }
    
    // 마지막 시점의 실제 위치도 gps_log[k-1]여야 함
    int answer = dp[k - 1][gps_log[k - 1]];
    
    return answer == INF ? -1 : answer;
}
